docs/design-proposals/node/Unified-Vertical-Scaling-for-Container-and-VM-Pods.md

```diff
@@ -319,3 +341,129 @@ Other components:
 * check how the change of meaning of resource requests influence other
   Kubernetes components.
 
+### Changes in Arktos-vm-runtime-service (CRI for VM workload types)
+
+#### Specific for VM workload
+
+1. Hotplug(hotUnplug) vs Memory Ballooning driver for Memory resizing
+
+   Currently per our knowledge these are two technologies that supports live memory resizing for VMs. Both have been there for
+   Linux for years, and both has pros and cons in supporting memory resizing for VMs. There are also some proposals to combine
+   them to better support VM resizing for Linux VMs.
+
+   In Arktos VM runtime, it is proposed to use hotplug(hotUnplug) in current release for the following reasons:
+
+    1. Numa support
+    2. Less memory fragmentation
+    3. Better control from VM spec patching
+
+2. VM memory live update
+
+    Both attach and detach device operations are done via the AttachDeviceFlag and DetachDeviceFlag API, with flag to
+    affect the running VM and persist the changes to the domain configuration for next reboot as well.
+       
+       AttachDeviceFlags(memoryDeviceDefinition, libvirt.DOMAIN_DEVICE_MODIFY_CONFIG|libvirt.DOMAIN_DEVICE_MODIFY_LIVE)
+       DetachDeviceFlags(memoryDeviceDefinition, libvirt.DOMAIN_DEVICE_MODIFY_CONFIG|libvirt.DOMAIN_DEVICE_MODIFY_LIVE)
+   
+3. Handle the async operation for memory unplug
+
+   Add a handler for the VIR_DOMAIN_EVENT_ID_DEVICE_REMOVED (future works). Current design returns in the UpdateContainerResource()
+   once after the DetachDeviceFlag() returns success without checking if the memory device is removed at the VM configuration in
+   the guest. It is done such way mainly for not holding the UpdateContainerResource() API for unexpected longer time due to
+   memory unplug.
+   
+4. Some limitations/constraints for VM workload types
+
+    1. Memory device size: 
+       Currently it is single size, with 128MiB each. i.e. if a VM patch increases the memory as 1Gi, 8 memory devices 
+       will be attached to the VM. The operations will still in one UpdateContainerResource() call though. i.e. Arktos VM 
+       runtime service will handle the attach operation internally with the calculation of memory increment in the patch.
+       
+    2. Max memory scale-up value
+       Currently a VM memory slots are set to 16 in Arktos VM runtime, so 16 * 128 Mi, which is 2Gi max scale-up value.
+       This is assumed the host has plenty allocatable memory.
+       
+    3. Guest OS Kernel requirements
+       Guest OS kernel must support device hotPlug and hotUnplug.
```

Comment:
Is it Linux only? For example, will a Windows VM pod eligible for vertical scaling?

Reply:
this is a TBD for requrements for the vm support in general. currently it is Linux that we are mainly testing on.

Reply:
Arktos will support windows VM. will add this back to our goal to support vertical scaling as well.

Reply:
Windows support is not there today (CRI API needs update), but it can be added fairly quickly. The upstream version is awaiting Patrick's review and overall Windows needs to add support for UpdateContainerResources - last I heard from patrick, they were looking to migrate to containerd, will take some time, and we decided not to block alpha on that.